Firestore/core/src/firebase/firestore/local/memory_lru_reference_delegate.cc

```diff
@@ -0,0 +1,190 @@
+#include "Firestore/core/src/firebase/firestore/local/lru_garbage_collector.h"
+#include "Firestore/core/src/firebase/firestore/local/memory_mutation_queue.h"
+#include "Firestore/core/src/firebase/firestore/local/memory_persistence.h"
+#include "Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h"
+#include "Firestore/core/src/firebase/firestore/local/reference_set.h"
+#include "Firestore/core/src/firebase/firestore/local/remote_document_cache.h"
+#include "Firestore/core/src/firebase/firestore/local/sizer.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+using model::DocumentKey;
+using model::ListenSequenceNumber;
+
+MemoryLruReferenceDelegate::MemoryLruReferenceDelegate(
+    MemoryPersistence* persistence,
+    LruParams lru_params,
+    std::unique_ptr<Sizer> sizer)
+    : persistence_(persistence), sizer_(std::move(sizer)) {
+  gc_ = absl::make_unique<LruGarbageCollector>(this, lru_params);
+
+  // Theoretically this is always 0, since this is all in-memory...
+  ListenSequenceNumber highest_sequence_number =
+      persistence_->query_cache()->highest_listen_sequence_number();
+  listen_sequence_ = absl::make_unique<ListenSequence>(highest_sequence_number);
+}
+
+LruGarbageCollector* MemoryLruReferenceDelegate::garbage_collector() {
+  return gc_.get();
+}
+
+ListenSequenceNumber MemoryLruReferenceDelegate::current_sequence_number() {
+  HARD_ASSERT(current_sequence_number_ != kListenSequenceNumberInvalid,
+              "Asking for a sequence number outside of a transaction");
+  return current_sequence_number_;
+}
+
+void MemoryLruReferenceDelegate::AddInMemoryPins(ReferenceSet* set) {
+  // Technically can't assert this, due to restartWithNoopGarbageCollector (for
+  // now...) FSTAssert(additional_references_ == nil, @"Overwriting additional
+  // references");
+  additional_references_ = set;
+}
+
+void MemoryLruReferenceDelegate::RemoveTarget(const QueryData& queryData) {
+  QueryData updated =
+      queryData.Copy(queryData.snapshot_version(), queryData.resume_token(),
+                     current_sequence_number_);
+  persistence_->query_cache()->UpdateTarget(updated);
+}
+
+void MemoryLruReferenceDelegate::UpdateLimboDocument(
+    const model::DocumentKey& key) {
+  sequence_numbers_[key] = current_sequence_number_;
+}
+
+void MemoryLruReferenceDelegate::OnTransactionStarted(absl::string_view label) {
+  current_sequence_number_ = listen_sequence_->Next();
+}
+
+void MemoryLruReferenceDelegate::OnTransactionCommitted() {
+  current_sequence_number_ = kListenSequenceNumberInvalid;
+}
+
+void MemoryLruReferenceDelegate::EnumerateTargets(
+    const TargetCallback& callback) {
+  return persistence_->query_cache()->EnumerateTargets(callback);
+}
+
+void MemoryLruReferenceDelegate::EnumerateOrphanedDocuments(
+    const OrphanedDocumentCallback& callback) {
+  for (const auto& entry : sequence_numbers_) {
+    ListenSequenceNumber sequenceNumber = entry.second;
```

Comment:
Optional: it's a little weird that `entry.second` is accessed before `entry.first`.